### ISSUE TITLE
backup option: implement custom naming for files ( ansible_managed like )

### ISSUE THREAD

**Issue by andreagrax** [ansible/ansible#82408]
### Summary

This feature requests is based on  [issue 16305](https://github.com/ansible/ansible/issues/16305) but takes out of the way  the management of the permissions for the new files that would be created, since we are just adding more control over the naming scheme (except for #3, maybe )
It would be great if the naming of the backup file can be controlled via something similar to the [ansible_managed](https://docs.ansible.com/archive/ansible/2.4/intro_configuration.html#ansible-managed) inside the `ansible.cfg` 

in this way setting it I could address the following use cases:
1. decide to create a single backup file ( setting it to `${filename}.backup` ) 
2. change the format of the backup name  ( `$filename.$( date "+%s")`)
3.  change the path of the file ( `.backup/$filename.$( date "+%s")` . **Creation of the backup directory must be handled by another task** 

### Issue Type

Feature Idea

### Component Name

template, copy

### Additional Information

ansible.cfg 
```cfg
[defaults]
ansible_backup = {target_file}.%Y%m%d_%H%M%S
```

task.yml
```yaml 
- name: Apply the template
  hosts: your_hosts
  vars:
    dest_file: /path/to/your/file.conf

  tasks:
    - name: Apply the template
      template:
        src: your_template.j2
        dest: "{{ dest_file }}"
        backup: true
```

This will create `/path/to/your/file.conf.20231212.142835`

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

**Comment by ansibot** [comment]
Files identified in the description:

* [`lib/ansible/plugins/lookup/template.py`](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/lookup/template.py)
* [`lib/ansible/plugins/action/copy.py`](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/copy.py)
* [`lib/ansible/plugins/action/template.py`](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/template.py)
* [`lib/ansible/modules/template.py`](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/template.py)
* [`lib/ansible/modules/copy.py`](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/copy.py)

If these files are incorrect, please update the `component name` section of the description or use the [component](https://github.com/ansible/ansibotmini#commands) bot command.

**Comment by nightgryphon** [comment]
Any chances this will be implemented any soon? 
File naming with PID before timestamp completely mess the backup files order so that's become a pain to find the correct backup file to restore and produce human errors

**Comment by nkfilis** [comment]
> Any chances this will be implemented any soon? File naming with PID before timestamp completely mess the backup files order so that's become a pain to find the correct backup file to restore and produce human errors

probably not. there's a Pull Request to simply reverse the order of PID and timestamp: https://github.com/ansible/ansible/pull/82434 but it was closed instead of just merged. So no relief in sight apart from maintaining a private patch

**Comment by bcoca** [comment]
check PR linked above and test if this meets your needs

**Comment by andreagrax** [comment]
hello @bcoca , sorry for the late reply.

I am testing with this playbook :

```yaml
---
- name: Default empty playbook
  hosts: localhost
  become: false
  gather_facts: true

  tasks:

  - name: Copy using inline content
    ansible.builtin.copy:
      content: '{{ ansible_date_time.iso8601 }}'
      dest: /tmp/template
      backup: true

  - name: Copy using inline content
    ansible.builtin.copy:
      content: '{{ ansible_date_time.iso8601 }}'
      dest: /tmp/template
      backup: true
```


and this command from your branch :

```bash
[root@ebd71149a926 ~]# ANSIBLE_BACKUP_FILE_NAME_TEMPLATE='{basename}.{timestamp}.{pid}'  ANSIBLE_LOCALHOST_WARNING=False ANSIBLE_INVENTORY_UNPARSED_WARNING=False python3.11  ./ansible/bin/ansible-playbook  -e ansible_connection=local -i localhost, do.yml    ; ls -l /tmp/template*
[WARNING]: You are running the development version of Ansible. You should only run Ansible from "devel" if you are modifying the Ansible engine, or trying out
features under development. This is a rapidly changing source of code and can become unstable at any point.

PLAY [Default empty playbook] **********************************************************************************************************************************

TASK [Gathering Facts] *****************************************************************************************************************************************
[WARNING]: Platform linux on host localhost is using the discovered Python interpreter at /usr/bin/python3.11, but future installation of another Python
interpreter could change the meaning of that path. See https://docs.ansible.com/ansible-core/devel/reference_appendices/interpreter_discovery.html for more
information.
ok: [localhost]

TASK [Copy using inline content] *******************************************************************************************************************************
changed: [localhost]

TASK [Copy using inline content] *******************************************************************************************************************************
ok: [localhost]

PLAY RECAP *****************************************************************************************************************************************************
localhost                  : ok=3    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0

-rw-r--r-- 1 root root 20 Mar  6 13:43 /tmp/template
-rw-r--r-- 1 root root 20 Mar  6 13:39 /tmp/template.1357.2025-03-06@13:43:17~
-rw-r--r-- 1 root root 20 Mar  6 13:43 /tmp/template.1458.2025-03-06@13:43:28~
-rw-r--r-- 1 root root 20 Mar  6 13:43 /tmp/template.1559.2025-03-06@13:43:36~
-rw-r--r-- 1 root root 20 Mar  6 13:43 /tmp/template.1660.2025-03-06@13:43:39~
```


But I guess that I'm setting wrong `ANSIBLE_BACKUP_FILE_NAME_TEMPLATE`

**Comment by nkfilis** [comment]
try `{basename}.{timestamp}.{pid}` without whitespace.
Having an `ansible.cfg` like this
```
[defaults]
backup_file_name_template='{basename}.{timestamp}.{pid}'
```
works for me

**Comment by andreagrax** [comment]
tried with the `ansible.cfg`  , new parameter got recognized, but getting an error while running the playbook


# checking version and remote

    [root@7bb950d9d7aa ansible]# git rev-parse --short HEAD  ; git rev-parse --abbrev-ref HEAD ; git remote -v
    8a40bef120
    backup_template
    origin  https://github.com/bcoca/ansible.git (fetch)
    origin  https://github.com/bcoca/ansible.git (push)
    
# show config file
    [root@ed4d224055d2 ~]# cat $ANSIBLE_CONFIG
    [defaults]
    backup_file_name_template='{basename}.{timestamp}.{pid}.mytest'
    
# show config (BACKUP_FILE_NAME_TEMPLATE seems ok)

    [root@7bb950d9d7aa ramdisk]# ANSIBLE_CONFIG=/root/ansible.cfg python3.11 /mnt/ramdisk/ansible/bin/ansible-config   dump  --only-changed
    [WARNING]: You are running the development version of Ansible. You should only run Ansible from "devel" if you are modifying the Ansible engine, or trying out
    features under development. This is a rapidly changing source of code and can become unstable at any point.
    BACKUP_FILE_NAME_TEMPLATE(/root/ansible.cfg) = {basename}.{timestamp}.{pid}.mytest
    CONFIG_FILE() = /root/ansible.cfg
    DEFAULT_PRIVATE_KEY_FILE(env: ANSIBLE_PRIVATE_KEY_FILE) = /root/.ssh/id_ed25519_tpsreans
    DEFAULT_REMOTE_USER(env: ANSIBLE_REMOTE_USER) = tpsreans
    DEFAULT_VAULT_PASSWORD_FILE(env: ANSIBLE_VAULT_PASSWORD_FILE) = /tmp/ansible_vault_pw
    DISPLAY_SKIPPED_HOSTS(env: ANSIBLE_DISPLAY_SKIPPED_HOSTS) = True
    HOST_KEY_CHECKING(env: ANSIBLE_HOST_KEY_CHECKING) = False
    

# show playbook
    [root@ed4d224055d2 ~]# cat /mnt/ramdisk/do.yml
    ---
    - name: Default empty playbook
      hosts: localhost
      become: false
      gather_facts: true
    
      tasks:
        - name: Copy using inline content
          ansible.builtin.copy:
            content: '{{ ansible_date_time.iso8601 }}'
            dest: /tmp/template
            backup: true
    
        - name: Copy using inline content
          ansible.builtin.copy:
            content: '{{ ansible_date_time.iso8601 }}'
            dest: /tmp/template
            backup: true

# run playbook

    [root@7bb950d9d7aa ramdisk]# ANSIBLE_CONFIG=/root/ansible.cfg python3.11 /mnt/ramdisk/ansible/bin/ansible-playbook -e ansible_connection=local  /mnt/ramdisk/do.yml  -l localhost
    [WARNING]: You are running the development version of Ansible. You should only run Ansible from "devel" if you are modifying the Ansible engine, or trying out
    features under development. This is a rapidly changing source of code and can become unstable at any point.
    [WARNING]: No inventory was parsed, only implicit localhost is available
    [WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'
    [WARNING]: errors were encountered during the plugin load for ansible.builtin.copy: ["expected 'except' or 'finally' block (helpers.py, line 39)"]
    
    PLAY [Default empty playbook] **********************************************************************************************************************************
    ERROR! Unexpected Exception, this is probably a bug: expected 'except' or 'finally' block (helpers.py, line 39)
    to see the full traceback, use -vvv
    [root@7bb950d9d7aa ramdisk]# ANSIBLE_CONFIG=/root/ansible.cfg python3.11 /mnt/ramdisk/ansible/bin/ansible-playbook -e ansible_connection=local  /mnt/ramdisk/do2.yml  -l localhost -vvv
    [WARNING]: You are running the development version of Ansible. You should only run Ansible from "devel" if you are modifying the Ansible engine, or trying out
    features under development. This is a rapidly changing source of code and can become unstable at any point.
    ansible-playbook [core 2.19.0.dev0] (backup_template 8a40bef120) last updated 2025/03/07 06:25:57 (GMT +000)
      config file = /root/ansible.cfg
      configured module search path = ['/root/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
      ansible python module location = /mnt/ramdisk/ansible/lib/ansible
      ansible collection location = /root/.ansible/collections:/usr/share/ansible/collections
      executable location = /mnt/ramdisk/ansible/bin/ansible-playbook
      python version = 3.11.7 (main, Oct  9 2024, 00:00:00) [GCC 11.4.1 20231218 (Red Hat 11.4.1-3)] (/usr/bin/python3.11)
      jinja version = 3.1.4
      libyaml = True
    Using /root/ansible.cfg as config file
    host_list declined parsing /etc/ansible/hosts as it did not pass its verify_file() method
    Skipping due to inventory source not existing or not being readable by the current user
    script declined parsing /etc/ansible/hosts as it did not pass its verify_file() method
    auto declined parsing /etc/ansible/hosts as it did not pass its verify_file() method
    Skipping due to inventory source not existing or not being readable by the current user
    yaml declined parsing /etc/ansible/hosts as it did not pass its verify_file() method
    Skipping due to inventory source not existing or not being readable by the current user
    ini declined parsing /etc/ansible/hosts as it did not pass its verify_file() method
    Skipping due to inventory source not existing or not being readable by the current user
    toml declined parsing /etc/ansible/hosts as it did not pass its verify_file() method
    [WARNING]: No inventory was parsed, only implicit localhost is available
    [WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'
    [WARNING]: errors were encountered during the plugin load for ansible.builtin.copy: ["expected 'except' or 'finally' block (helpers.py, line 39)"]
    Skipping callback 'default', as we already have a stdout callback.
    Skipping callback 'minimal', as we already have a stdout callback.
    Skipping callback 'oneline', as we already have a stdout callback.
    
    PLAYBOOK: do2.yml **********************************************************************************************************************************************
    1 plays in /mnt/ramdisk/do2.yml
    
    PLAY [Default empty playbook] **********************************************************************************************************************************
    ERROR! Unexpected Exception, this is probably a bug: expected 'except' or 'finally' block (helpers.py, line 39)
    the full traceback was:
    
    Traceback (most recent call last):
      File "/mnt/ramdisk/ansible/lib/ansible/cli/__init__.py", line 656, in cli_executor
        exit_code = cli.run()
                    ^^^^^^^^^
      File "/mnt/ramdisk/ansible/bin/ansible-playbook", line 151, in run
        results = pbex.run()
                  ^^^^^^^^^^
      File "/mnt/ramdisk/ansible/lib/ansible/executor/playbook_executor.py", line 188, in run
        result = self._tqm.run(play=play)
                 ^^^^^^^^^^^^^^^^^^^^^^^^
      File "/mnt/ramdisk/ansible/lib/ansible/executor/task_queue_manager.py", line 344, in run
        play_return = strategy.run(iterator, play_context)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/mnt/ramdisk/ansible/lib/ansible/plugins/strategy/linear.py", line 147, in run
        action = action_loader.get(task_action, class_only=True, collection_list=task.collections)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/mnt/ramdisk/ansible/lib/ansible/plugins/loader.py", line 860, in get
        return self.get_with_context(name, *args, **kwargs).object
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/mnt/ramdisk/ansible/lib/ansible/plugins/loader.py", line 878, in get_with_context
        plugin_load_context = self.find_plugin_with_context(name, collection_list=collection_list)
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/mnt/ramdisk/ansible/lib/ansible/plugins/loader.py", line 591, in find_plugin_with_context
        result = self._resolve_plugin_step(name, mod_type, ignore_deprecated, check_aliases, collection_list, plugin_load_context=plugin_load_context)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/mnt/ramdisk/ansible/lib/ansible/plugins/loader.py", line 687, in _resolve_plugin_step
        return self._find_plugin_legacy(name, plugin_load_context, ignore_deprecated, check_aliases, suffix)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/mnt/ramdisk/ansible/lib/ansible/plugins/loader.py", line 718, in _find_plugin_legacy
        for path_with_context in (p for p in self._get_paths_with_context() if p.path not in self._searched_paths and os.path.isdir(to_bytes(p.path))):
                                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/mnt/ramdisk/ansible/lib/ansible/plugins/loader.py", line 371, in _get_paths_with_context
        ret.extend([PluginPathContext(p, True) for p in self._get_package_paths(subdirs=subdirs)])
                                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/mnt/ramdisk/ansible/lib/ansible/plugins/loader.py", line 333, in _get_package_paths
        m = __import__(self.package)
            ^^^^^^^^^^^^^^^^^^^^^^^^
      File "/mnt/ramdisk/ansible/lib/ansible/plugins/action/__init__.py", line 21, in <module>
        from ansible.config.helpers import get_validated_backup_file_name_template
      File "/mnt/ramdisk/ansible/lib/ansible/config/helpers.py", line 39
        if errors:
        ^^
    SyntaxError: expected 'except' or 'finally' block

**Comment by bcoca** [comment]
i did a bunch of iterations after i asked you to test and then went on PTO, so PR was 'brokenish' .. I'll get back to it soon

**Comment by andreagrax** [comment]
re-run the test . still not creating with new name.

# repo info 

    [root@d1e1a662bf75 ansible]# git branch ; git remote -v ; git  rev-parse --short HEAD
    * backup_template
      devel
    origin  https://github.com/bcoca/ansible.git (fetch)
    origin  https://github.com/bcoca/ansible.git (push)
    39e755e

# ansible cfg

    [root@d1e1a662bf75 ansible]# cat /mnt/ramdisk/ansible.cfg
    [defaults]
    backup_file_name_template='{basename}.{timestamp}.{pid}.mytest'



# run playbook 

    [root@d1e1a662bf75 ansible]#  ANSIBLE_CONFIG=./ansible.cfg python3.11 /mnt/ramdisk/ansible/bin/ansible-playbook  -e ansible_connection=local  /mnt/ramdisk/do.yml --diff -v
    No config file found; using defaults

    PLAY [Default empty playbook] ****************************************************************************************************************************************************************

    TASK [Gathering Facts] ***********************************************************************************************************************************************************************
    ok: [localhost]

    TASK [Copy using inline content] *************************************************************************************************************************************************************
    --- before: /tmp/template
    +++ after: /root/.ansible/tmp/ansible-local-15775ib3ns1u/tmp22z3v2sz
    @@ -1 +1 @@
    -2025-05-06T14:51:22Z
    \ No newline at end of file
    +2025-05-06T14:51:49Z
    \ No newline at end of file

    changed: [localhost] => {"backup_file": "/tmp/template.1670.2025-05-06@14:51:50~", "changed": true, "checksum": "5b4be3aa1ee784ee61f3e712e827c553242ade84", "dest": "/tmp/template", "gid": 0, "group": "root", "md5sum": "7c5f6a8da30e85ff4d238a0b6200d102", "mode": "0644", "owner": "root", "size": 20, "src": "/root/.ansible/tmp/ansible-tmp-1746543109.8438168-1644-74749086215654/source", "state": "file", "uid": 0}

    TASK [Copy using inline content] *************************************************************************************************************************************************************
    ok: [localhost] => {"changed": false, "checksum": "5b4be3aa1ee784ee61f3e712e827c553242ade84", "dest": "/tmp/template", "gid": 0, "group": "root", "mode": "0644", "owner": "root", "path": "/tmp/template", "size": 20, "state": "file", "uid": 0}

    PLAY RECAP ***********************************************************************************************************************************************************************************
    localhost                  : ok=3    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
